enhancements/network/bgp-overview.md

```diff
@@ -0,0 +1,264 @@
+---
+title: bgp-overview
+authors:
+  - "@russellb"
+reviewers:
+  - "@danwinship"
+  - "@squeed"
+approvers:
+  - TBD
+creation-date: 2020-08-11
+last-updated: 2020-08-11
+status: informational
+see-also:
+  - https://github.com/openshift/enhancements/pull/356
+  - https://github.com/openshift/enhancements/pull/394
+---
+
+# OpenShift and BGP
+
+BGP (Border Gateway Protocol) is a dynamic routing protocol used to power a
+number of important networking use cases, including routing in the global
+internet.  For some helpful resources on learning more about BGP, see the
+[BGP Resources](#bgp-resources) section of this document.
+
+The purpose of this document is to serve as a high level overview of how BGP
+relates to OpenShift.  We will review a number of use cases, discuss how they
+could be supported, and link to existing related work.  Any future enhancements
+related to BGP should be able to rely on this one to provide some broader
+context and should continue to update this document as necessary.
+
+## Use Cases
+
+### L3 (Layer 3) Network Fabric
+
+The use case here does not require any changes to OpenShift.  It is still
+useful to call out as it is one use of BGP that is relevant to OpenShift,
+particularly with on premise clusters.
+
+It is possible to use BGP between the routers used to build the L3 network that
+OpenShift Nodes are connected to.
+
+### L3 Redundancy for Nodes
+
+*First, an aside on terminology: This is referring to IP Routing, and
+not the OpenShift feature called Routing that works at the HTTP layer.  There
+are also OpenShift features that use the word "egress", but in this context, we
+only mean to refer to traffic that is leaving the cluster and not related to
+any other features that use that name.*
+
+Some networks may have more than one router accessible from Nodes for
+availability reasons.  In this case an administrator may want to use BGP down
+to the Nodes themselves so they are aware of the multiple routing options and
+potentially use ECMP (Equal Cost Multi Path) to distribute traffic
+across each of the routers.
+
+Another motivator for this configuration is to provide L3 redundancy instead of
+using bonding to provide L2 redundancy.  A node would publish a route to its IP
+to two different routers accessible by two different network interfaces. In
+this way
+
+This capability is not provided by OpenShift today.  It may be possible as a
+custom configuration, but not with all network providers.  With OpenShift-SDN,
+it should be possible to run your own BGP routing daemon on the host via a
+`DaemonSet`.
+
+OVN-Kubernetes uses a different host network configuration where egress
+traffic is routed by Open vSwitch flows programmed by OVN.  It means that a
+dynamic routing daemon must also feed those routes into OVN configuration, so
+it's not as straight forward as just running an existing daemon on the host.
```

Comment:
On the OpenStack side we have been considering a configuration where OVN would create a /32 or /128 loopback IP inside a namespace. FRR would leak those host routes to the default VRF for exporting to the ToR routing switch. In this case the BGP daemon might only accept default routes from peers and would only export routes to /32 or /128 IPs. Egress traffic to IPs that are not on the same compute host would be handed to the kernel for outbound routing. Ingress traffic would be forwarded to OVN to be delivered to the instance via flows. This reduces complexity because then the dynamic routing daemon does not have to feed routes back into the OVN configuration.